### ISSUE TITLE
Handling multiple in-flight requests

### ISSUE THREAD

**Issue by carver** [sigp/discv5#198]
When sending multiple requests, the current design queues them, waiting for a response to the first request before sending the next one:

https://github.com/sigp/discv5/blob/a9ded0490229418cb61e747d55d3e0879b62e1d7/src/handler/mod.rs#L468-L481

If a client has a 100ms roundtrip time with a peer, that puts a hard upper limit of 10 messages per second with that peer. That limits the "bandwidth" to less than 12.8kB/s (102 kb/s).

There's nothing apparent in the protocol spec that requires 1 message at a time. Including the request ID even suggests that it is intended to support multiple in-flight requests.

What are maintainers' thoughts on reviewing a PR that would support multiple in-flight requests?

**Comment by fjl** [comment]
It is certainly possible to run concurrent requests, but please note that it will make session handling much more complicated. Remember that in discv5, it is assumed that the session may be dropped at any moment. If the node has multiple requests in flight with a peer, and the peer answers with a challenge (WHOAREYOU), all pending requests would need to be replayed. It is much easier to implement the handshake flow when there is the guarantee that only a single request is pending for each peer.

**Comment by AgeManning** [comment]
Yep. I think in its original design it allowed for multiple con-current in-flight requests. But handling all the edge cases got pretty unwieldy and there were significant code simplifications we can make by having a single one per-peer. 

The other issue you may run into, is that we also need a way to determine whether a peer is still alive/online. There is a configuration parameter that allows you to retry a request if it fails. If we have many outgoing requests at a time and some fail, should we consider the connection failed?

At the moment I think the default is that if we get a request time out, we just consider the peer offline and move on. This is helpful for discovery queries where we are asking many different peers lots of things at once, rather than a single peer a lot of things.

**Comment by njgheorghita** [comment]
Took a stab at implementing this in #200 - would love to hear some feedback. Did I take the right approach? Did I miss any edge cases? 

> If we have many outgoing requests at a time and some fail, should we consider the connection failed?

I took the approach of "yes". As well as replaying all active requests if a new challenge is received.

I had very successful results using this inside `trin` to facilitate our `utp` transfers, which were really struggling beforehand when concurrent requests were not allowed.

**Comment by ackintosh** [comment]
It has been implemented by #200 and was released in [v0.4.0](https://github.com/sigp/discv5/releases/tag/v0.4.0). 🎉
